tests/check_framework/test_caches.py

```diff
@@ -28,3 +33,93 @@ def test_default_cache_not_included(self):
+                'LOCATION': Path.cwd() / 'fake',
+            }
+        },
+        MEDIA_ROOT=Path.cwd(),
+    )
+    def test_exposed_on_media_root_path(self):
+        self.assertEqual(check_cache_location_not_exposed(None), [Warning(
+            "Your configuration might expose your cache or lead to corruption of your data.",
+            hint="Your cache 'default' is inside your 'MEDIA_ROOT'",
+            id='cache.W002',
+        )])
+
+    @override_settings(
+        CACHES={
+            'default': {
+                'BACKEND': 'django.core.cache.backends.filebased.FileBasedCache',
+                'LOCATION': Path.cwd() / 'fake2',
+            }
+        },
+        MEDIA_ROOT=Path.cwd() / 'fake',
+    )
+    def test_no_exposed_on_media_root_path(self):
+        self.assertEqual(check_cache_location_not_exposed(None), [])
+
+    @override_settings(
+        CACHES={
+            'default': {
+                'BACKEND': 'django.core.cache.backends.filebased.FileBasedCache',
+                'LOCATION': Path.cwd() / 'fake',
+            }
+        },
+        STATIC_ROOT=Path.cwd(),
+    )
+    def test_exposed_on_static_root(self):
+        self.assertEqual(check_cache_location_not_exposed(None), [Warning(
+            "Your configuration might expose your cache or lead to corruption of your data.",
+            hint="Your cache 'default' is inside your 'STATIC_ROOT'",
+            id='cache.W002',
+        )])
+
+    @override_settings(
+        CACHES={
+            'default': {
+                'BACKEND': 'django.core.cache.backends.filebased.FileBasedCache',
+                'LOCATION': Path.cwd() / 'fake1',
+            }
+        },
+        STATIC_ROOT=Path.cwd() / 'fake2',
+    )
+    def test_no_exposed_on_static_root(self):
+        self.assertEqual(check_cache_location_not_exposed(None), [])
+
+    @override_settings(
+        CACHES={
+            'default': {
+                'BACKEND': 'django.core.cache.backends.filebased.FileBasedCache',
+                'LOCATION': Path.cwd() / 'fake',
+            }
+        },
+        STATICFILES_DIRS=[
+            Path.cwd(),
+        ],
+    )
+    def test_exposed_on_staticfiles_dirs(self):
+        self.assertEqual(check_cache_location_not_exposed(None), [Warning(
+            "Your configuration might expose your cache or lead to corruption of your data.",
+            hint="Your cache 'default' is inside your 'STATICFILES_DIRS'",
+            id='cache.W002',
+        )])
+
+    @override_settings(
+        CACHES={
+            'default': {
+                'BACKEND': 'django.core.cache.backends.filebased.FileBasedCache',
+                'LOCATION': Path.cwd() / 'fake1',
+            }
+        },
+        STATICFILES_DIRS=[
+            Path.cwd() / 'fake2',
+        ],
+    )
+    def test_no_exposed_on_staticfiles_dirs(self):
+        self.assertEqual(check_cache_location_not_exposed(None), [])
```

Comment:
We can then test all the different combinations and use `self.subTest(...)` to reduce duplication as follows:

```suggestion


class CheckCacheLocationTest(SimpleTestCase):
    @staticmethod
    def get_settings(name, cache_path, other_path):
        return {
            'CACHES': {
                'default': {
                    'BACKEND': 'django.core.cache.backends.filebased.FileBasedCache',
                    'LOCATION': cache_path,
                },
            },
            name: [other_path] if name == 'STATICFILES_DIRS' else other_path,
        }

    def assertWarningWithHint(self, hint):
        self.assertEqual(check_cache_location_not_exposed(None), [Warning(
            'Your configuration might expose your cache or lead to corruption of your data.',
            hint=hint,
            id='cache.W002',
        )])

    def test_cache_path_matches_other_path(self):
        root = pathlib.Path.cwd()
        for name in ('MEDIA_ROOT', 'STATIC_ROOT', 'STATICFILES_DIRS'):
            settings = self.get_settings(name, root, root)
            with self.subTest(setting=name), override_settings(**settings):
                self.assertWarningWithHint(f"Your 'default' cache path matches your {name}.")

    def test_cache_path_inside_other_path(self):
        root = pathlib.Path.cwd()
        for name in ('MEDIA_ROOT', 'STATIC_ROOT', 'STATICFILES_DIRS'):
            settings = self.get_settings(name, root / 'cache', root)
            with self.subTest(setting=name), override_settings(**settings):
                self.assertWarningWithHint(f"Your 'default' cache path is inside your {name}.")

    def test_cache_path_contains_other_path(self):
        root = pathlib.Path.cwd()
        for name in ('MEDIA_ROOT', 'STATIC_ROOT', 'STATICFILES_DIRS'):
            settings = self.get_settings(name, root, root / 'other')
            with self.subTest(setting=name), override_settings(**settings):
                self.assertWarningWithHint(f"Your 'default' cache path contains your {name}.")

    def test_cache_path_does_not_conflict_with_other_path(self):
        root = pathlib.Path.cwd()
        for name in ('MEDIA_ROOT', 'STATIC_ROOT', 'STATICFILES_DIRS'):
            settings = self.get_settings(name, root / 'cache', root / 'other')
            with self.subTest(setting=name), override_settings(**settings):
                self.assertEqual(check_cache_location_not_exposed(None), [])
```